### ISSUE TITLE
Early symbol deallocation in FunctionalTests_testBlocksWorldOperatorSubgoaling

### ISSUE THREAD

**Issue by bluechill** [SoarGroup/Soar#229]
SVS reuses a freed pointer.  Couldn't obviously figure out why.

Output:

```
AddressSanitizer debugger support is active. Memory error breakpoint has been installed and you can now use the 'memory history' command.
NAME        PASS   STOP   NOTIFY
==========  =====  =====  ======
SIGPIPE     true   false  true 
NAME        PASS   STOP   NOTIFY
==========  =====  =====  ======
SIGPIPE     true   false  true 

testBlocksWorldOperatorSubgoaling: =================================================================
==20612==ERROR: AddressSanitizer: heap-use-after-free on address 0x61100152e340 at pc 0x000101d83b71 bp 0x7000000773f0 sp 0x7000000773e8
READ of size 8 at 0x61100152e340 thread T37
.    #0 0x101d83b70 in soar_module::remove_module_wme(agent_struct*, wme_struct*) soar_module.cpp:61
    #1 0x1020d754d in soar_interface::remove_wme(wme_struct*) soar_interface.cpp:86
    #2 0x101caa1af in sgwme::~sgwme() svs.cpp:64
    #3 0x101cac814 in sgwme::~sgwme() svs.cpp:57
    #4 0x101cacb18 in sgwme::node_update(sgnode*, sgnode::change_type, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) svs.cpp:102
    #5 0x1017e3bb1 in sgnode::send_update(sgnode::change_type, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) sgnode.cpp:152
    #6 0x1017f6f68 in sgnode::send_update(sgnode::change_type) sgnode.h:113
    #7 0x1017df585 in sgnode::~sgnode() sgnode.cpp:33
    #8 0x1017e7ba8 in group_node::~group_node() sgnode.cpp:296
    #9 0x1017e7c74 in group_node::~group_node() sgnode.cpp:289
    #10 0x1017e7c98 in group_node::~group_node() sgnode.cpp:289
    #11 0x102507441 in scene::~scene() scene.cpp:90
    #12 0x1025074f4 in scene::~scene() scene.cpp:88
    #13 0x102507548 in scene::~scene() scene.cpp:88
    #14 0x101caff11 in svs_state::~svs_state() svs.cpp:198
    #15 0x101cb00e4 in svs_state::~svs_state() svs.cpp:187
    #16 0x101cb0108 in svs_state::~svs_state() svs.cpp:187
    #17 0x101cbf0eb in svs::state_deletion_callback(symbol_struct*) svs.cpp:415
    #18 0x10170dab5 in remove_existing_context_and_descendents(agent_struct*, symbol_struct*) decide.cpp:2879
    #19 0x101d3d74e in remove_operator_if_necessary(agent_struct*, slot_struct*, wme_struct*) consistency.cpp:107
    #20 0x1016e6b9a in do_acceptable_preference_wme_changes_for_slot(agent_struct*, slot_struct*) decide.cpp:215
    #21 0x1016e7c3f in do_buffered_acceptable_preference_wme_changes(agent_struct*) decide.cpp:279
    #22 0x101715184 in do_buffered_wm_and_ownership_changes(agent_struct*) decide.cpp:3400
    #23 0x10171539c in do_working_memory_phase(agent_struct*) decide.cpp:3431
    #24 0x1021c8ea9 in do_one_top_level_phase(agent_struct*) init_soar.cpp:917
    #25 0x1021cdc67 in run_for_n_decision_cycles(agent_struct*, long long) init_soar.cpp:1479
    #26 0x1015cca47 in sml::AgentSML::Step(sml::smlRunStepSize) sml_AgentSML.cpp:507
    #27 0x1015cc799 in sml::AgentSML::StepInClientThread(sml::smlRunStepSize) sml_AgentSML.cpp:453
    #28 0x101d7fc70 in sml::RunScheduler::RunScheduledAgents(bool, sml::smlRunStepSize, unsigned long long, sml::smlRunFlags, sml::smlRunStepSize, bool) sml_RunScheduler.cpp:843
    #29 0x1021fe6c4 in sml::KernelSML::DirectRun(char const*, bool, int, int, unsigned long long) sml_KernelSML.cpp:800
    #30 0x101dcbce4 in sml::EmbeddedConnection::DirectRun(char const*, bool, int, int, unsigned long long) sml_EmbeddedConnection.cpp:275
    #31 0x101bbc3f9 in sml::Agent::RunSelf(int, sml::smlRunStepSize) sml_ClientAgent.cpp:1218
    #32 0x10013d433 in FunctionalTestHarness::runTestExecute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:45
    #33 0x10013fbeb in FunctionalTestHarness::runTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:67
    #34 0x100193ebe in FunctionalTests::testBlocksWorldOperatorSubgoaling() FunctionalTests.cpp:48
    #35 0x1001ee4ee in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&>(std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&&&) __functional_base:382
    #36 0x1001ee0bb in std::__1::__function::__func<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>, std::__1::allocator<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*> >, void ()>::operator()() functional:1407
    #37 0x10009bcac in std::__1::function<void ()>::operator()() const functional:1793
    #38 0x10009af59 in TestRunner::run() TestRunner.cpp:28
    #39 0x1002a3fbc in void* std::__1::__thread_proxy<std::__1::tuple<void (TestRunner::*)(), TestRunner*> >(void*) __functional_base:382
    #40 0x102e58a50 in _pthread_body (/usr/lib/system/introspection/libsystem_pthread.dylib+0x3a50)
    #41 0x102e589cd in _pthread_start (/usr/lib/system/introspection/libsystem_pthread.dylib+0x39cd)
    #42 0x102e56018 in thread_start (/usr/lib/system/introspection/libsystem_pthread.dylib+0x1018)

0x61100152e340 is located 0 bytes inside of 208-byte region [0x61100152e340,0x61100152e410)
freed by thread T37 here:
    #0 0x100429b19 in wrap_free (/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/7.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib+0x42b19)
    #1 0x1017dd89c in void Memory_Manager::free_with_pool<wme_struct>(MemoryPoolType, wme_struct*) memory_manager.h:257
    #2 0x1017dd1b9 in deallocate_wme(agent_struct*, wme_struct*) wmem.cpp:330
    #3 0x1017dd84c in wme_remove_ref(agent_struct*, wme_struct*) wmem.h:175
    #4 0x1017dcf57 in do_buffered_wm_changes(agent_struct*) wmem.cpp:309
    #5 0x101715196 in do_buffered_wm_and_ownership_changes(agent_struct*) decide.cpp:3402
    #6 0x10171539c in do_working_memory_phase(agent_struct*) decide.cpp:3431
    #7 0x1021c8ea9 in do_one_top_level_phase(agent_struct*) init_soar.cpp:917
    #8 0x1021cdc67 in run_for_n_decision_cycles(agent_struct*, long long) init_soar.cpp:1479
    #9 0x1015cca47 in sml::AgentSML::Step(sml::smlRunStepSize) sml_AgentSML.cpp:507
    #10 0x1015cc799 in sml::AgentSML::StepInClientThread(sml::smlRunStepSize) sml_AgentSML.cpp:453
    #11 0x101d7fc70 in sml::RunScheduler::RunScheduledAgents(bool, sml::smlRunStepSize, unsigned long long, sml::smlRunFlags, sml::smlRunStepSize, bool) sml_RunScheduler.cpp:843
    #12 0x1021fe6c4 in sml::KernelSML::DirectRun(char const*, bool, int, int, unsigned long long) sml_KernelSML.cpp:800
    #13 0x101dcbce4 in sml::EmbeddedConnection::DirectRun(char const*, bool, int, int, unsigned long long) sml_EmbeddedConnection.cpp:275
    #14 0x101bbc3f9 in sml::Agent::RunSelf(int, sml::smlRunStepSize) sml_ClientAgent.cpp:1218
    #15 0x10013d433 in FunctionalTestHarness::runTestExecute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:45
    #16 0x10013fbeb in FunctionalTestHarness::runTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:67
    #17 0x100193ebe in FunctionalTests::testBlocksWorldOperatorSubgoaling() FunctionalTests.cpp:48
    #18 0x1001ee4ee in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&>(std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&&&) __functional_base:382
    #19 0x1001ee0bb in std::__1::__function::__func<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>, std::__1::allocator<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*> >, void ()>::operator()() functional:1407
    #20 0x10009bcac in std::__1::function<void ()>::operator()() const functional:1793
    #21 0x10009af59 in TestRunner::run() TestRunner.cpp:28
    #22 0x1002a3fbc in void* std::__1::__thread_proxy<std::__1::tuple<void (TestRunner::*)(), TestRunner*> >(void*) __functional_base:382
    #23 0x102e58a50 in _pthread_body (/usr/lib/system/introspection/libsystem_pthread.dylib+0x3a50)
    #24 0x102e589cd in _pthread_start (/usr/lib/system/introspection/libsystem_pthread.dylib+0x39cd)
    #25 0x102e56018 in thread_start (/usr/lib/system/introspection/libsystem_pthread.dylib+0x1018)

previously allocated by thread T37 here:
    #0 0x100429950 in wrap_malloc (/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/7.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib+0x42950)
    #1 0x1017dd44f in void Memory_Manager::allocate_with_pool<wme_struct>(MemoryPoolType, wme_struct**) memory_manager.h:197
    #2 0x1017daf16 in make_wme(agent_struct*, symbol_struct*, symbol_struct*, symbol_struct*, bool) wmem.cpp:84
    #3 0x101d83937 in soar_module::add_module_wme(agent_struct*, symbol_struct*, symbol_struct*, symbol_struct*) soar_module.cpp:51
    #4 0x1020d9bbd in soar_interface::make_wme(symbol_struct*, symbol_struct*, symbol_struct*) soar_interface.cpp:170
    #5 0x101cc65c4 in wme_struct* soar_interface::make_wme<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(symbol_struct*, symbol_struct*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) soar_interface.h:117
    #6 0x101ca7407 in sgwme::sgwme(soar_interface*, symbol_struct*, sgwme*, sgnode*) svs.cpp:37
    #7 0x101ca90c4 in sgwme::sgwme(soar_interface*, symbol_struct*, sgwme*, sgnode*) svs.cpp:35
    #8 0x101cae3a1 in svs_state::init() svs.cpp:224
    #9 0x101caed7a in svs_state::svs_state(symbol_struct*, svs_state*) svs.cpp:183
    #10 0x101caee44 in svs_state::svs_state(symbol_struct*, svs_state*) svs.cpp:181
    #11 0x101cbe6e7 in svs::state_creation_callback(symbol_struct*) svs.cpp:397
    #12 0x101711833 in create_new_context(agent_struct*, symbol_struct*, unsigned char) decide.cpp:3013
    #13 0x10171498a in decide_context_slot(agent_struct*, symbol_struct*, slot_struct*, bool) decide.cpp:3294
    #14 0x101714feb in decide_context_slots(agent_struct*, bool) decide.cpp:3357
    #15 0x1017153e8 in do_decision_phase(agent_struct*, bool) decide.cpp:3438
    #16 0x1021cc244 in do_one_top_level_phase(agent_struct*) init_soar.cpp:1208
    #17 0x1021cdc67 in run_for_n_decision_cycles(agent_struct*, long long) init_soar.cpp:1479
    #18 0x1015cca47 in sml::AgentSML::Step(sml::smlRunStepSize) sml_AgentSML.cpp:507
    #19 0x1015cc799 in sml::AgentSML::StepInClientThread(sml::smlRunStepSize) sml_AgentSML.cpp:453
    #20 0x101d7fc70 in sml::RunScheduler::RunScheduledAgents(bool, sml::smlRunStepSize, unsigned long long, sml::smlRunFlags, sml::smlRunStepSize, bool) sml_RunScheduler.cpp:843
    #21 0x1021fe6c4 in sml::KernelSML::DirectRun(char const*, bool, int, int, unsigned long long) sml_KernelSML.cpp:800
    #22 0x101dcbce4 in sml::EmbeddedConnection::DirectRun(char const*, bool, int, int, unsigned long long) sml_EmbeddedConnection.cpp:275
    #23 0x101bbc3f9 in sml::Agent::RunSelf(int, sml::smlRunStepSize) sml_ClientAgent.cpp:1218
    #24 0x10013d433 in FunctionalTestHarness::runTestExecute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:45
    #25 0x10013fbeb in FunctionalTestHarness::runTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) FunctionalTestHarness.cpp:67
    #26 0x100193ebe in FunctionalTests::testBlocksWorldOperatorSubgoaling() FunctionalTests.cpp:48
    #27 0x1001ee4ee in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&>(std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>&&&) __functional_base:382
    #28 0x1001ee0bb in std::__1::__function::__func<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*>, std::__1::allocator<std::__1::__bind<void (FunctionalTests::*)(), FunctionalTests*> >, void ()>::operator()() functional:1407
    #29 0x10009bcac in std::__1::function<void ()>::operator()() const functional:1793

Thread T37 created by T0 here:
    #0 0x10041ec99 in wrap_pthread_create (/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/7.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib+0x37c99)
    #1 0x1002a33a2 in std::__1::thread::thread<void (TestRunner::*)(), TestRunner*&, void>(void (TestRunner::*&&)(), TestRunner*&&&) thread:359
    #2 0x1001cff84 in std::__1::thread::thread<void (TestRunner::*)(), TestRunner*&, void>(void (TestRunner::*&&)(), TestRunner*&&&) thread:355
    #3 0x1001cb109 in main testMain.cpp:109
    #4 0x7fff9e4db5ac in start (/usr/lib/system/libdyld.dylib+0x35ac)
    #5 0x0 in 0x0
 (/Users/bluechill/Library/Developer/Xcode/DerivedData/Soar-fbrkwlvdiyteohhfnthtdujfkeih/Build/Products/Debug/JSoar-UnitTests+0x0)

SUMMARY: AddressSanitizer: heap-use-after-free soar_module.cpp:61 soar_module::remove_module_wme(agent_struct*, wme_struct*)
Shadow bytes around the buggy address:
  0x1c22002a5c10: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x1c22002a5c20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c22002a5c30: fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c22002a5c40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c22002a5c50: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
=>0x1c22002a5c60: fa fa fa fa fa fa fa fa[fd]fd fd fd fd fd fd fd
  0x1c22002a5c70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c22002a5c80: fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c22002a5c90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c22002a5ca0: 00 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa
  0x1c22002a5cb0: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==20612==ABORTING
Warning: hit breakpoint while running function, skipping commands and conditions to prevent recursion.AddressSanitizer report breakpoint hit. Use 'thread info -s' to get extended information about the report.
```

**Comment by mazina** [comment]
Confirmed this still exists, though my backtrace is different than yours.

**Comment by mazina** [comment]
Fixed in 9.6.0
